### PR TITLE
etc.c: give every NaN an identity of its own

### PR DESCRIPTION
**Pull request by takumin**
## Summary

A NaN is equal to no value, its own operand included, so `==` can never find one. A container searching for the NaN it holds has nothing left but the object, and mruby gave it nothing to hold on to: every NaN was normalized to one representation, so a NaN was either the same object as every other NaN or, under `MRB_NO_BOXING`, not even the same object as itself.

```ruby
z = [0.0][0]; a = z / z; b = z / z   # two NaNs, made apart
```

## Changes

Five commits, each green on its own. The first four are the PRs this one is built on, and this PR shrinks to its last commit once they land:

| commit | PR |
| --- | --- |
| `array.c: take an Array element for equal to itself in #==` | #7359 |
| `mruby-array-ext: walk an Array in C for include? and member?` | #7360 |
| `mruby-enum-ext: count an Array in C` | #7361 |
| `object.c: compare what a Float holds bit for bit under MRB_NO_BOXING` | #7362 |
| `etc.c: give every NaN an identity of its own` | this one |

The four below it are what a NaN needs before an identity is worth giving it: the searches an array makes asking for the object, and an `equal?` that reads the same thing in every boxing. Each is worth having on its own and none of them mentions a NaN in its own right.

| File | What |
| --- | --- |
| `include/mruby/value.h` | adds `MRB_NAN_SERIAL_MAX`, `MRB_NAN_QUIET_BIT`, `mrb_float_bits` and `mrb_nan_serialize()`, for the boxings that put the count in the Float |
| `include/mruby.h` | adds `mrb_state.nan_serial` |
| `include/mruby/boxing_nan.h` | the count goes in the payload where a Float is made |
| `include/mruby/boxing_no.h` | the same, in `SET_FLOAT_VALUE` |
| `include/mruby/boxing_word.h` | a heap Float is a Float at every width now, the one exception having gone |
| `src/etc.c` | adds `mrb_nan_serial_next()`; word boxing hands every NaN to the heap and drops the sentinel it kept for one; `mrb_obj_id()` gives a heap NaN the id of the object it is |
| `src/hash.c` | a key that is not equal to itself is asked for the object |
| `test/t/float.rb` | `equal?`, `object_id`, `index`, `Array#==` and Hash lookup, for one NaN and for two made apart |
| `mrbgems/mruby-array-ext/test/array.rb` | `uniq`, `-`, `&` and `include?` with a NaN |
| `mrbgems/mruby-enum-ext/test/enum.rb` | `count` with a NaN |
| `mrbgems/mruby-pack/test/pack.rb` | a NaN that signals is read back as a NaN, and two reads are two objects |

The assertions are asked where the method they ask about is answered: `include?` and `count` are `mruby-array-ext`'s and `mruby-enum-ext`'s, so a build without those gems is not asked about them.

### Where the identity comes from

Every NaN made is now one of its own, and what carries that is whatever the boxing has room for.

- **word boxing** gives a NaN the heap, where the object it is answers for it and two objects are never one. A word that is not an immediate already points at an object, so no reader is asked anything new, and this is what CRuby does as well: a NaN falls outside the flonum range and is allocated, as an infinity is. The 32-bit `MRB_USE_FLOAT32` build already took this path for want of room in its word; it is the rule at every width now, which leaves the sentinel range at the four floats that are equal to themselves and drops the count that range used to carry.
- **nan boxing** and **no boxing** keep a Float in the value, so there is no object to be had and the NaN has to carry what tells it from another. That goes in the payload, which no arithmetic reads: a count, `mrb_state.nan_serial`, given every bit the payload can spare. That is **51 bits** under `MRB_NO_BOXING` and **48** under `MRB_NAN_BOXING`, where the tag saying which value a word holds begins.
- The count is written with the quiet bit set, which is what keeps every pattern it makes a NaN. A NaN that signals can have every bit it is set inside the field the count takes, and clearing those for a count of zero would leave the exponent alone, which is an infinity.

A count is a bound and an object is not, so the two differ in what a long run can reach. Holding one NaN while 2^51 more are made leaves two that hold the same bits; at the 50M NaNs a second this machine manages, that is over a year of making nothing else. **`MRB_USE_FLOAT32` under `MRB_NO_BOXING` is the one build where the bound is a near one**: a binary32 NaN has 22 payload bits and no more, so 4,194,304 NaNs bring the count round. There is no object to fall back on there, a Float being the value itself, and no other build is affected: `MRB_NAN_BOXING` refuses `MRB_USE_FLOAT32`, and word boxing keeps no count at all.

### What had to stop asking `==` alone

A **Hash** asks for the object where a key is not equal to itself, which only a NaN is, so the key it stored is found again. What an ordinary key pays is that comparison, made against the value already in hand.

**`mrb_obj_id()`** names what `equal?` compares, so the two have to agree. Under word boxing a Float that is not immediate is an object, and its id came off the bits it holds; every NaN holds the same bits, so two that `equal?` now tells apart would have shared an id. A NaN is given the id of the object it is, which is what the same function returns for everything else on the heap. The other boxings need nothing here: the count is in the payload, so the bits already differ.

The other callers are the commits below this one: `mrb_obj_eq()` reading a Float bit for bit under `MRB_NO_BOXING`, and `Array#==`, `#include?`, `#member?` and `#count` comparing a pair with `mrb_equal()`, as `#index` and `#delete` already did. Each is worth having on its own, and together they are what makes a NaN's identity reach an answer once it has one.

## Behaviour

Every row answers as CRuby does after the change, and the three boxings answer alike, which they did not before. Bold marks where a build differs from CRuby today.

| expression | word / nan boxing | `MRB_NO_BOXING` | this PR | CRuby |
| --- | --- | --- | --- | --- |
| `a.equal?(a)` | true | **false** | true | true |
| `a.equal?(b)` | **true** | false | false | false |
| `a.object_id == c.object_id` | true | true | true | true |
| `a.object_id == b.object_id` | **true** | **true** | false | false |
| `[a].index(a)` | 0 | **nil** | 0 | 0 |
| `[a].index(b)` | **0** | nil | nil | nil |
| `[a].include?(a)` | **false** | **false** | true | true |
| `[a].include?(b)` | false | false | false | false |
| `[a].count(a)` | **0** | **0** | 1 | 1 |
| `[a].count(b)` | 0 | 0 | 0 | 0 |
| `[1.0, a] == [1.0, a]` | **false** | **false** | true | true |
| `[1.0, a] == [1.0, b]` | false | false | false | false |
| `[a, a].uniq.size` | 1 | **2** | 1 | 1 |
| `[a, b].uniq.size` | **1** | 2 | 2 | 2 |
| `{a => 1}[a]` | **nil** | **nil** | 1 | 1 |
| `{a => 1}[b]` | nil | nil | nil | nil |
| `([a] - [b]).size` | **0** | 1 | 1 | 1 |
| `([a] & [a]).size` | 1 | **0** | 1 | 1 |
| `pzero.equal?(nzero)` | false | **true** | false | false |

Measured against CRuby 4.0.6, with `a` and `b` built at run time as `z / z` from `z = [0.0][0]`, and `pzero` and `nzero` as `z + 0.0` and `z * -1.0`. The `this PR` column is the same in every one of the twelve boxing builds listed under Testing.

## Speed

```ruby
t = 0.0; i = 0
while i < 1000000 do t += i * 0.5; i += 1 end           # Float arithmetic

h = {}; i = 0
while i < 100000 do h[i * 0.5] = i; i += 1 end          # Float-keyed Hash
i = 0
while i < 100000 do h[i * 0.5]; i += 1 end

z = [0.0][0]; i = 0
while i < 100000 do z / z; i += 1 end                   # NaNs made
```

Best of 11 interleaved runs against master, with a master-vs-master control to read the noise by, and the instruction counts callgrind reads beside them:

| benchmark | master | this PR | control | Ir |
| --- | --- | --- | --- | --- |
| `t += i * 0.5`, 1M turns | 377 ms | 394 ms (1.05x) | -0.1% | 523,999,972 -> 523,999,972 |
| Float-keyed Hash, 100k insert + 100k lookup | 42.5 ms | 43.0 ms (1.01x) | -0.2% | +0.43% |
| 100k NaNs made in a loop | 6.0 ms | **8.2 ms (1.37x)** | +1.7% | +45.8% |

The Float path is untouched and the first row says so: the same instruction count to the last one over a million turns, which is the same code either way; the wall clock difference beside it is layout, not work. What the Hash pays is the comparison a key now makes against itself, +0.43%.

The third row is what this costs. A NaN takes an object under word boxing now, and the loop pays about 174 instructions for each one it makes. That is the trade the design makes, and it is the one CRuby makes too: a NaN is not a flonum there either. Nothing pays it but a program making NaNs, which is a program on an error path.

The wall clock times were taken with the default build, which carries all four commits; the two arriving from the PRs below do not touch any of these three loops, and their instruction counts against master are identical.

## Size

`bin/mruby` `.text` for every build in `build_config/ci/gcc-clang.rb`, for all five commits:

| build | master | this PR | delta |
| --- | --- | --- | --- |
| `ascii-ctype` | 1,293,718 | 1,294,198 | +480 |
| `bintest` | 1,307,110 | 1,307,590 | +480 |
| `byte-string` | 1,271,494 | 1,271,974 | +480 |
| `cxx_abi` | 1,334,041 | 1,334,601 | +560 |
| `full-debug` (-O0) | 1,913,478 | 1,914,182 | +704 |

In `bintest` five objects add up to the +480 exactly, and **208 of it is this commit**:

| object | delta | commit |
| --- | --- | --- |
| `src/array.o` | **-80** | `Array#==` |
| `mruby-array-ext/src/array.o` | +144 | `include?` / `member?` |
| `mruby-enum-ext/src/enum.o` | +208 | `count` |
| `src/hash.o` | +240 | this one |
| `src/etc.o` | **-32** | this one |

`etc.o` shrinks on balance: the sentinel range word boxing kept for a NaN and the count that range carried are gone, against the test `mrb_obj_id()` now makes. `object.o` is unchanged, its edit being under `MRB_NO_BOXING`.

## Testing

`rake test` passes for the default build, for `build_config/host-nofloat.rb`, and for every build in `build_config/ci/gcc-clang.rb` including the C++ ABI one, with no compiler warnings. The one `-Wmaybe-uninitialized` about `eq` in `hash.c` that a `MRB_NO_BOXING` build emits is master's, unmoved by the edit beside it.

Because the change is one the boxing decides, it was also run over twelve builds of the full-core gembox:

- 64-bit: `MRB_NO_BOXING`, `MRB_WORD_BOXING` and `MRB_NAN_BOXING`, and the first two again with `MRB_USE_FLOAT32` (which nan boxing refuses).
- 32-bit: the same three, `MRB_USE_FLOAT32` with `MRB_INT32` under the first two, and `MRB_WORDBOX_NO_INLINE_FLOAT` at both float widths, which covers the Float that never sits in a word.

Seven of the twelve pass outright. The five `MRB_USE_FLOAT32` builds fail exactly what they fail on master: three assertions about how many digits a binary32 prints (`Float#to_s`, `mrb_vformat` and one in mruby-sprintf) and one crash, none of them about a NaN, and the counts are the same before and after this change. Each of the five commits was built and run on its own, so none of them is red in the middle.

The tests added build their two NaNs at run time from a variable so that nothing folds them into a single literal, and assert the same answers in every boxing, which is the point of the change. The one in `mrbgems/mruby-pack/test/pack.rb` is there rather than in `test/t/` because a NaN that signals cannot be spelled without `unpack`: it reads such a pattern back and asks that it be a NaN and that two reads be two objects.

## Environment

<details>
<summary>Machine, toolchain and the compile lines these numbers were taken with</summary>

| | |
| --- | --- |
| OS | Ubuntu 24.04.4 LTS |
| Kernel | Linux 7.0.0-30-generic x86_64 |
| CPU | AMD Ryzen 9 5950X (16 cores) |
| C compiler | gcc (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0 |
| 32-bit compiler | i686-linux-gnu-gcc 13.3.0 |
| binutils | GNU ld (GNU Binutils) 2.47.20260726 |
| valgrind | valgrind-3.27.1 (callgrind) |
| CRuby | 4.0.6 (2026-07-14 revision 03b6d3f889) |

The wall clock and instruction counts were taken with the default build:

```console
gcc -std=gnu99 -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DMRB_USE_COMPLEX -DMRB_USE_BIGINT -DMRB_USE_DEBUG_HOOK
```

The `## Size` table was taken with `build_config/ci/gcc-clang.rb`:

```console
bintest      gcc -std=gnu99 -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -DMRB_GC_FIXED_ARENA -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER -DMRB_USE_DEBUG_HOOK
ascii-ctype  gcc -std=gnu99 -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -DMRB_USE_ASCII_CTYPE -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER
byte-string  gcc -std=gnu99 -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER
cxx_abi      gcc -g -O3 -Wall -Wundef -Wwrite-strings -x c++ -std=gnu++03 -DMRB_GC_FIXED_ARENA -DMRB_USE_CXX_EXCEPTION -DMRB_USE_CXX_ABI -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER
full-debug   gcc -std=gnu99 -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -g3 -O0 -DMRB_GC_STRESS -DMRB_USE_DEBUG_HOOK -DMRB_DEBUG -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER
```

`cxx_abi` compiles mruby as C++ with `gcc -x c++ -std=gnu++03`; g++ only links. `full-debug` is the one build at `-O0`, `enable_debug` putting `-g3 -O0` after the toolchain default of `-g -O3`.

</details>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of `NaN` values so each independently created value maintains distinct object identity.
  * Updated array and hash operations to correctly recognize the exact `NaN` instance being searched for.
  * Preserved signaling `NaN` values when unpacking floating-point data.
  * Improved `NaN` recognition and behavior across supported floating-point formats.
* **Tests**
  * Added coverage for `NaN` comparisons, collection operations, identity-based lookups, and floating-point edge cases.

### CODE REVIEW

File: src/etc.c [hunk 1]
```diff
@@ -291,9 +310,18 @@ mrb_word_boxing_float_value(mrb_state *mrb, mrb_float f)
...
-  v.w = 0;
-  v.f = f;
-  v.w = (v.w<<2) | 2;
+  if (f != f) {
+    /* a NaN is the object it is, and a word holding the whole float has
+       nothing left to tell one from another with */
+    v.p = mrb_obj_alloc(mrb, MRB_TT_FLOAT, mrb->float_class);
+    mrb_rfloat_set(v.fp, f);
+    v.bp->frozen = 1;
+  }
```
[the remark this answers is not in the document]
Comment (takumin) [reply src/etc.c:325]: Reproduced, and fixed in 965c60d4.

Under word boxing a NaN is a heap object after this change, but `mrb_obj_id()` read the bits the Float holds, and every NaN holds the same ones. Two NaNs that `equal?` had just been taught to tell apart still shared an id:

| | `a.equal?(b)` | `a.object_id == b.object_id` |
| --- | --- | --- |
| master | true | true |
| before the fix | false | **true** |
| after the fix | false | false |

`mrb_obj_id()` now hands a heap NaN the id of the object it is, which is what the same function returns for everything else that is not immediate. The other two boxings needed nothing: the count sits in the payload there, so the bits of two NaNs already differ and `mrb_float_id()` was telling them apart before the fix as well.

Checked across all twelve boxing builds listed under Testing; `equal?` and `object_id` agree in every one of them. `test/t/float.rb` now asserts both directions, that a NaN passed around keeps its id and that two made apart do not share one.

File: mrbgems/mruby-array-ext/src/array.c [hunk 2]
```diff
@@ -1777,6 +1777,43 @@ ary_min(mrb_state *mrb, mrb_value self)
...
+
+     Nothing accumulates in the GC arena over the walk, which is why there is
+     no arena restore in it, as there is none in `Array#index`: what a call
+     leaves behind is its return value, and this walk returns at the first one
+     that is true. The answers it walks past are false, which is immediate. */
+  for (mrb_int i = 0; i < RARRAY_LEN(self); i++) {
+    if (mrb_equal(mrb, RARRAY_PTR(self)[i], obj)) {
+      return mrb_true_value();
+    }
+  }
```
[the remark this answers is not in the document]
Comment (takumin) [reply mrbgems/mruby-array-ext/src/array.c:1813]: I do not think this one holds, and the reason is that what a `==` allocates does not survive the call it was allocated in.

`mrb_funcall_with_block()` saves the arena index on entry and restores it on the way out, then puts the return value through `mrb_gc_protect()`:

```c
/* src/vm.c */
int ai = mrb_gc_arena_save(mrb);
...
mrb_gc_arena_restore(mrb, ai);
mrb_gc_protect(mrb, val);
return val;
```

So a false comparison leaves the arena exactly as it found it, whatever it built along the way. What it can leave is its return value, and `mrb_gc_protect()` takes no immediate:

```c
/* src/gc.c */
mrb_gc_protect(mrb_state *mrb, mrb_value obj)
{
  if (mrb_immediate_p(obj)) return;
  ...
```

`false` and `nil` are the only falsy values in Ruby and both are immediate, so a walk that keeps going past a false answer consumes no arena at all. The walk stops at the first answer that is not false, so it never carries a protected value into another turn either. That is the whole of the difference from `Array#count`, which goes past a true answer and does need the restore it now has in #7361.

Measured, with an `==` allocating hard on every call and answering false, over 1000 elements under `MRB_GC_FIXED_ARENA`:

```ruby
class Allocating
  def ==(other)
    10.times { |i| "garbage-#{i}" * 20 }
    [1, 2, 3].map { |x| { x => "v" } }
    false
  end
end
a = Array.new(1000) { Allocating.new }
a.include?(:missing)   # false
a.member?(:missing)    # false
a.index(:missing)      # nil
a.count(:missing)      # 0
```

All four answer, on master and here alike. `Array#index` in core has searched without an arena restore for the same reason, and adding one to this walk would cost an instruction per element to prevent something that cannot happen. The comment above the loop says as much, so the next reader does not have to work it out again.
